core/src/main/java/hudson/model/Job.java

```diff
@@ -679,7 +671,56 @@ public void movedTo(DirectlyModifiableTopLevelItemGroup destination, AbstractIte
         }
     }
 
-    @Override public void delete() throws IOException, InterruptedException {
+    @Override
+    public void delete() throws IOException, InterruptedException {
+        checkPermission(DELETE);
+        // if a build is in progress. Cancel it.
+        if (this instanceof Queue.Task) {
+            // clear any items in the queue so they do not get picket up
+            Queue.getInstance().cancel((Queue.Task) this);
+            // interrupt any builds in progress
+            Map<Executor, Queue.Executable> building = new LinkedHashMap<>();
+            for (Computer c : Jenkins.getInstance().getComputers()) {
+                for (Executor e : c.getOneOffExecutors()) {
+                    WorkUnit workUnit = e.getCurrentWorkUnit();
+                    if (workUnit != null && (workUnit.work == this || workUnit.work.getOwnerTask() == this)) {
+                        building.put(e, e.getCurrentExecutable());
+                        e.interrupt(Result.ABORTED);
+                    }
+                }
+                for (Executor e : c.getExecutors()) {
+                    WorkUnit workUnit = e.getCurrentWorkUnit();
+                    if (workUnit != null && (workUnit.work == this || workUnit.work.getOwnerTask() == this)) {
+                        building.put(e, e.getCurrentExecutable());
+                        e.interrupt(Result.ABORTED);
```

Comment:
Better to factor into a helper method. Better yet, introduce `Computer.getAllExecutors()`, since that is a common problem. (See `Executor.of` etc.)